quaddtype/numpy_quaddtype/src/umath.cpp

```diff
@@ -0,0 +1,815 @@
+        }
+    }
+
+    // Set up output descriptor
+    if (given_descrs[2] == NULL) {
+        loop_descrs[2] = (PyArray_Descr *)new_quaddtype_instance(target_backend);
+        if (!loop_descrs[2]) {
+            return (NPY_CASTING)-1;
+        }
+    }
+    else {
+        QuadPrecDTypeObject *descr_out = (QuadPrecDTypeObject *)given_descrs[2];
+        if (descr_out->backend != target_backend) {
+            loop_descrs[2] = (PyArray_Descr *)new_quaddtype_instance(target_backend);
+            if (!loop_descrs[2]) {
+                return (NPY_CASTING)-1;
+            }
+        }
+        else {
+            Py_INCREF(given_descrs[2]);
+            loop_descrs[2] = given_descrs[2];
+        }
+    }
+    return casting;
+}
+
+template <binary_op_quad_def sleef_op, binary_op_longdouble_def longdouble_op>
+int
+quad_generic_binop_strided_loop_unaligned(PyArrayMethod_Context *context, char *const data[],
+                                          npy_intp const dimensions[], npy_intp const strides[],
+                                          NpyAuxData *auxdata)
+{
+    npy_intp N = dimensions[0];
+    char *in1_ptr = data[0], *in2_ptr = data[1];
+    char *out_ptr = data[2];
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp out_stride = strides[2];
+
+    QuadPrecDTypeObject *descr = (QuadPrecDTypeObject *)context->descriptors[0];
+    QuadBackendType backend = descr->backend;
+    size_t elem_size = (backend == BACKEND_SLEEF) ? sizeof(Sleef_quad) : sizeof(long double);
+
+    quad_value in1, in2, out;
+    while (N--) {
+        memcpy(&in1, in1_ptr, elem_size);
+        memcpy(&in2, in2_ptr, elem_size);
+        if (backend == BACKEND_SLEEF) {
+            sleef_op(&out.sleef_value, &in1.sleef_value, &in2.sleef_value);
+        }
+        else {
+            longdouble_op(&out.longdouble_value, &in1.longdouble_value, &in2.longdouble_value);
+        }
+        memcpy(out_ptr, &out, elem_size);
+
+        in1_ptr += in1_stride;
+        in2_ptr += in2_stride;
+        out_ptr += out_stride;
+    }
+    return 0;
+}
+
+template <binary_op_quad_def sleef_op, binary_op_longdouble_def longdouble_op>
+int
+quad_generic_binop_strided_loop_aligned(PyArrayMethod_Context *context, char *const data[],
+                                        npy_intp const dimensions[], npy_intp const strides[],
+                                        NpyAuxData *auxdata)
+{
+    npy_intp N = dimensions[0];
+    char *in1_ptr = data[0], *in2_ptr = data[1];
+    char *out_ptr = data[2];
+    npy_intp in1_stride = strides[0];
+    npy_intp in2_stride = strides[1];
+    npy_intp out_stride = strides[2];
+
+    QuadPrecDTypeObject *descr = (QuadPrecDTypeObject *)context->descriptors[0];
+    QuadBackendType backend = descr->backend;
+
+    while (N--) {
+        if (backend == BACKEND_SLEEF) {
+            sleef_op((Sleef_quad *)out_ptr, (Sleef_quad *)in1_ptr, (Sleef_quad *)in2_ptr);
```

Comment:
Here and in the unary op implementation, there's a chance that either or both `in1_ptr` and `in2_ptr` are the same location as `out_ptr`. Is that problematic? You should be able to test that by making sure that e.g. `np.multiply(a, b, out=a)` or `np.multiply(a, a, out=a)` is doing the right thing.

Reply:
Getting the expected results for both ops
```bash
In [14]: a = np.array([1,2,3], dtype=npq.QuadPrecDType())

In [15]: b = np.array([4,5,6], dtype=npq.QuadPrecDType())

In [16]: np.multiply(a, b, out=a)
Out[16]: 
array([QuadPrecision('4.0e+000', backend='sleef'),
       QuadPrecision('1.0e+001', backend='sleef'),
       QuadPrecision('1.8e+001', backend='sleef')],
      dtype=QuadPrecDType(backend='sleef'))

In [17]: a
Out[17]: 
array([QuadPrecision('4.0e+000', backend='sleef'),
       QuadPrecision('1.0e+001', backend='sleef'),
       QuadPrecision('1.8e+001', backend='sleef')],
      dtype=QuadPrecDType(backend='sleef'))

In [18]: np.multiply(a, a, out=a)
Out[18]: 
array([QuadPrecision('1.6e+001', backend='sleef'),
       QuadPrecision('1.0e+002', backend='sleef'),
       QuadPrecision('3.24e+002', backend='sleef')],
      dtype=QuadPrecDType(backend='sleef'))

In [19]: a
Out[19]: 
array([QuadPrecision('1.6e+001', backend='sleef'),
       QuadPrecision('1.0e+002', backend='sleef'),
       QuadPrecision('3.24e+002', backend='sleef')],
      dtype=QuadPrecDType(backend='sleef'))
```